### ISSUE TITLE
Rescued Albatrosses (Graveyard) get stuck in Peragenor

### ISSUE THREAD

**Issue by CodeKatta** [endless-sky/endless-sky#8781]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Describe the bug

After hunting some Ka'het in the Graveyard I rescued three Remnant ships (took those missions since I was there anyway).
With jumpdrives installed I saw no need to take the long route through the wormholes, so I jumped from _Gerenus_ (conveniently one of the Albatrosses was stranded there) to _Farinus_ to get back to Remnant space.
Usually this shouldn't be a problem and ships without jumpdrives find a different (if longer) route if the systems are connected.
The one Ibis I rescued appeared as expected after a few moments in _Farinus_, the two Albatrosses on the other hand never came even after waiting a few minutes in FFW-mode.
So I went the other way around (over _Edusa_) to see where they got lost. After I exited the wormhole in _Peragenor_ they immediately appeared on the radar, however, for some reason they darted towards the lower left corner and didn't make any effort to turn around.
I tried it a second time but without waiting. And although I arrived in _Peragenor_ before the rescued ships, the Albatrosses simply ignored me and headed in a southwesterly direction as soon as they emerged from hyperspace. The Ibis on the other hand flew towards my position and kept floating around my ship, as they always do when they catch up.

I don't know exactly what attracts them to the lower left corner and why only Albatrosses show this behavior but I guess this is not the way it's supposed to work. So far I had only Albatrosses and one Ibis, so I don't know if other types of ships are affected as well.

Originally, I noticed this on the [mobile version](https://f-droid.org/de/packages/com.github.thewierdnut.endless_mobile/) (0.10.0-33), so before complaining here, I downloaded the current Win64-[version](https://github.com/endless-sky/endless-sky/releases/download/v0.10.0/EndlessSky-win64-v0.10.0.zip) and checked it there as well. Haven't tested the Linux version yet but I guess it's the same and not platform related.

### Steps to Reproduce

1. Accept a Remnant rescue mission (those with 'distress') 
2. Go to the Graveyard and rescue those ships
3. Go to _Gerenus_ (you can also land on _Palarei_ before jumping to Remnant space)
4. Jump to _Farinus_ (or _Perfica_) and wait. Or go directly to _Peragenor_ to see what happens when they arrive.

### Expected Behavior

To arrive at my position after a certain time (as the Ibis did).

### Screenshots

![Endless Sky_Albatross_Peragenor](https://github.com/endless-sky/endless-sky/assets/26285588/b86f560c-fdac-42f6-b22e-cb345210dc20)

Here I arrived relatively early. If you wait a bit before getting to _Peragenor_ the distance can already be at seven digits.

### Operating System

Windows 10 Pro version 22H2 build 19045.2965

### Game Source

GitHub Releases

### Game Version

v0.10.0

**Comment by Quantumshark** [comment]
I've seen similar behaviour in freighter escorts in human space, I think we may have some kind of major AI bug somewhere.

**Comment by CodeKatta** [comment]
> I've seen similar behaviour in freighters from escort missions in human space, I think we may have some kind of major AI bug somewhere.

Sounds plausible, especially as it seems to be tied to a certain type of ship.

**Comment by Zitchas** [comment]
There's been a few instances of personalities (such as surveillance) just coasting into the void if they don't have anything to do, or if they can't quite nail the specific speed for their hyperjump, and stuff like that. I think it should be a bit improved in 0.10.1; but might not have fixed specifically what is going on here.

**Comment by UnorderedSigh** [comment]
I am having trouble reproducing both of your movement bugs, @CodeKatta, so I will need a save file to debug this further. (I'm using Linux at the top of master.)

**Comment by UnorderedSigh** [comment]
@CodeKatta - Could you please try #8858 and see if it fixes this problem? I have been unable to reproduce this one too.

**Comment by UnorderedSigh** [comment]
I have a branch with an adequite fix for this: if a ship is stranded, and has no helper, it'll revert to its normal behavior. Derelicts consider the player their helper, so they'll come to a stop waiting for help. This will also fix some of the problems @Zitchas mentioned:

> There's been a few instances of personalities (such as surveillance) just coasting into the void if they don't have anything to do, or if they can't quite nail the specific speed for their hyperjump, and stuff like that.

Those were caused by the `isStranded` checks. I changed it to `strandedWithoutHelper`

``` c++
  // Stranded ships that have a helper need to stop and be assisted.
 bool strandedWithHelper = isStranded &&
     (HasHelper(*it, isStranded) || it->GetPersonality().IsDerelict() || it->IsYours());
```

**Comment by UnorderedSigh** [comment]
Oh, forgot to mention: I need to test the branch a little bit more before opening a PR. I must make sure the new stranded logic works as intended. That'll require flying to some far-out systems where ships often run out of fuel (like the Hai-Syndicate corridor) and watching what happens.

**Comment by CodeKatta** [comment]
That's strange because the Ibis (and the three other ships) manage to reach my position in _Perfica_ or _Farinus_ just fine. Those other three ships (marked in blue) were also rescued, albeit not in the Graveyard.
If you look at those diagrams in the lower left corner you can see where they are and the bars for the different statuses and with the Ibis you can see the fuel bar also increasing significantly from time to time, so apparently it is able to recharge at a sun along the way. The question is why the Albatrosses aren't able to do the same.
And if they are stranded close to the system's sun they would recharge their fuel anyway so the 'stranded'-status would only be temporary since everything else (hull, etc.) is still at maximum.
Additionally, from _Peragenor_ to _Edusa_ you don't need fuel anyway since you have to take the wormhole so one would expect that they reach the 'stranded'-status there (however, _Edusa_ has a G-giant where you can refuel quite nicely if you have ramscoop-abilities, as all Remnant ships should have) so their predicament would also just be temporary.

An other question, if a ship is on its own and runs out of fuel in a system without an inhabited planet/station (or hostile, so it can't land) doesn't it automatically seek out the systems center so it can more efficiently refuel with the sun's solar wind? Or does it more or less stay where it jumped out of hyperspace?

**Comment by UnorderedSigh** [comment]
I know what happened because I traced the AI decisions for those ships. The ships decided they were stranded, and then did nothing due to an erroneous `continue` in AI::Step.

> An other question, if a ship is on its own and runs out of fuel in a system without an inhabited planet/station (or hostile, so it can't land) doesn't it automatically seek out the systems center so it can more efficiently refuel with the sun's solar wind? Or does it more or less stay where it jumped out of hyperspace?

Yes, they should. They didn't, though. However, the bug where a ship flies out infinitely fast has been fixed in the latest master. The ships just fly at their maximum speed.

The bad news is that I'm still seeing ships decide to do nothing. There's more debugging left to do.

**Comment by CodeKatta** [comment]
> I know what happened because I traced the AI decisions for those ships. The ships decided they were stranded, and then did nothing due to an erroneous `continue` in AI::Step.

Yes, I got that, I was just questioning the logic behind triggering the 'stranded'-status due to a lack of fuel as you can apparently recharge your fuel if you're not too far away from the system's sun(s). Those other four rescued ships somehow managed to do it on their own, just the two Albatrosses not.

> […] the ships will come to a stop waiting for you to refuel them. Unfortunately, they'll just stay there, stopped, until you fix them or someone blows them up.

But in that case you'd have to turn off their ramscoop-abilities or else that problem will solve itself over time as they can just collect the solar wind (which would make them no different from any other ship that ran out of fuel in that system).

> However, the bug where a ship flies out infinitely fast has been fixed in the latest master. The ships just fly at their maximum speed.

Or alternatively you could just have them play the Nyancat-theme as they dart off into the void 😁

> The bad news is that I'm still seeing ships decide to do nothing. There's more debugging left to do.

Oy, that can be quite tedious work. Good luck with that and thanks again.

**Comment by UnorderedSigh** [comment]
The next batch of movement bugs was related to ships that had a friendly target:

1. If the target left the system, it would not be de-targeted.
2. When the code decide to de-target, it didn't detarget. It just cleared a local copy of the target pointer, without clearing the actual target pointer in the ship.
3. If the ship had a friendly target, and did not move to scan it, then the AI was shut off; the ship did nothing.

**Comment by UnorderedSigh** [comment]
I think I've fixed all of the bugs that can cause the ship to move out into the void.

I found one other bug, which does not send ships to the void. Sometimes several ships will decide to scan each other, and get caught in the `CircleAround` behavior in `MoveIndependent`. Then they'll mutually circle around some location until they get bored and leave. Often, they're scanning members of their own fleet. I consider that to be a totally separate problem, so it belongs in a different PR. I'll open a PR for the "move out into the void" problem shortly.

**Comment by UnorderedSigh** [comment]
Also, only ships that have parents will move to the system center to refuel. Others will patrol from planet to planet. (Unless there is no planet in the system. I'm not sure what will happen then.) That, too, I consider a separate matter for a separate PR.

Edit: My prior post was wrong due to incorrect wording.

**Comment by UnorderedSigh** [comment]
> I, personally, would find an option to select (multiple) ships from the list to order them around quite useful.

Please open a separate issue for that. It is too unrelated to this one for your comment to be acted upon.

**Comment by CodeKatta** [comment]
> Please open a separate issue for that. It is too unrelated to this one for your comment to be acted upon.

Yes, would have done so anyway. Just thought since you're more familiar with the code you might've known if that's not intended behavior (a feature, so to say). In that case opening a separate issue would have been superfluous.

**Comment by UnorderedSigh** [comment]
Fix is in #8860

**Comment by CodeKatta** [comment]
> Also, only ships that have parents will move to the system center to refuel. Others will patrol from planet to planet. (Unless there is no planet in the system. I'm not sure what will happen then.) That, too, I consider a separate matter for a separate PR.

How much do ships 'know' about the systems they are about to enter? Are they 'aware' of any refueling options they have?
And if they do, wouldn't it be a sensible procedure to land on the last planet/station and refuel before entering empty/hostile space no matter how full your tanks already are?
But maybe they already do that (I'm not that familiar with what escort ships do when on their own).

**Comment by UnorderedSigh** [comment]
> How much do ships 'know' about the systems they are about to enter?

They know about the systems when planning their route. After that, they don't look at the next System.

> Are they 'aware' of any refueling options they have?

Once they are in a system, they know their refueling options.

> And if they do, wouldn't it be a sensible procedure to land on the last planet/station and refuel before entering empty/hostile space no matter how full your tanks already are?

The ship personalities restrict their options.

> But maybe they already do that (I'm not that familiar with what escort ships do when on their own).

Sometimes they do. The logic is quite complicated, and probably isn't 100% correct.

**Comment by CodeKatta** [comment]
> I don't know if this is an intended behavior.

OK, then maybe I open a new issue, although #7912 seems to have addressed the same topic (well, somehow).

**Comment by CodeKatta** [comment]
> They know about the systems when planning their route. After that, they don't look at the next System.

Ah OK, that's what I wanted to know. If they don't 'look ahead' once they're on their way then implementing a '_better safe than sorry_'-strategy (refueling even if your tanks are far from empty) is probably a bit more complicated.
